### ISSUE TITLE
Remove dependency on the Closure Library

### ISSUE THREAD

**Issue by tschaub** [openlayers/ol3#4128]
By moving away from the Closure Library, we free ourselves to write OpenLayers in a way that makes it more consumable for others.  Ideally, we could write ES6 modules.  We'll need to confirm that things will still work with the Compiler (if we need to transpile first, annotations will need to be preserved).

**Comment by bjornharrtell** [comment]
Interesting!

FYI I'm trying to move to ES6 modules in JSTS, from the Java source via AST transformation . Still just experimental, but one issue (https://github.com/bjornharrtell/java2estree/issues/13) I've seem to hit is that ES6 modules does not seem to support cyclic dependencies in class hierarchies i.e a subclass cannot inherit from a superclass that imports the subclass directly or indirectly.

This has to do with the module spec and more information about it can be read here:
https://github.com/babel/babel/issues/2061
http://www.2ality.com/2014/09/es6-modules-final.html

Because of this I'm currently considering using CommonJS module format instead of ES6 modules. I have not investigated if/where but I'm guessing there might be some of these cyclic dependencies in OL3.

**Comment by acanimal** [comment]
+1. 
Webpack can help with different module loading systems.

**Comment by gberaudo** [comment]
To give a bit of context, we discussed limitations of the build procedure
last week at the FOSS4G Seoul.

I wanted to allow users to easily compile several projects depending on OL3
together, without requiring the user to use
the closure compiler in it own application. For example, to be able to
compile Ngeo, OL3-Cesium and OL3 in one step and use the result as a
dependency.

The path to reach this goal was believed to be in standardizing the build
process, notably by shrinking or eliminating our dependency on the goog
library (not the compiler).
In the long run, it seemed that most of our needs can be handled by ES6
(standard inheritance, modules, ...) and transpilation from ES6 to ES5
(closure compiler).

As a side effect of this standardization, it was believed that improving
the build experience with closure would benefit all users.

**Comment by sweco-sebhar** [comment]
So is the project accepting PRs for any goog dependency removal from now on? Did some random checks and it seems goog.json (1 require occurence) and goog.string (12 require occurences) might be good candidates to work on. But perhaps it will be useful to create a script of some sorts that can make a top list of goog.* usage over the whole code base?

PS. Unfortunately it seems using CommonJS doesn't solve all issues that can occur with circular dependencies, at least as far as I understand these issues at this point. I still don't know if it will be a problem for OL3, but it if it is it's probably best to refactor away any problematic circular dependency if it comes to that.

**Comment by tschaub** [comment]
> So is the project accepting PRs for any goog dependency removal from now on?

Yes.  We'll leave `goog.provide` and `goog.require` until the rest of the library usage is gone.

**Comment by acanimal** [comment]
What solution do you want to use for module dependencies and loading?
I'm really happy (sorry for closure fans) with the removal of goog closure.
I hope I can contribute a bit to the project.

Cheers

|----------------------------------------------------------------
| The book of OpenLayers 3 - http://kcy.me/x29y
| OpenLayers Cookbook: http://kcy.me/ar2b
| Blog: http://www.acuriousanimal.com/blog
| Twitter: @acanimal
|----------------------------------------------------------------

**Comment by tschaub** [comment]
> What solution do you want to use for module dependencies and loading?

When we remove the Closure Library and use ES6 or CommonJS modules, people developing applications with OpenLayers will have more choices for a module loader - Browserify, Webpack, and jspm are obvious choices today (the choices will grow and native module support will eventually be an option).  What we choose to use for library development will matter less, since application developers won't have to use the same.

We haven't yet decided what will work best for library development.  The clear part is that the Closure Library limits how people can use OpenLayers for application development, and it makes sense to remove that impediment.

**Comment by probins** [comment]
> perhaps it will be useful to create a script of some sorts that can make a top list of goog.* usage over the whole code base?

`closurebuilder.py` can print out a dependency tree (the compiler has `output_manifest` which is similar), though I think the issue is more how dependent OL is on the library. `goog.json` you mention should be easy to convert to native, but `goog.math` for example might use code that will either have to be incorporated into OL or OL will have to use some other external lib.

**Comment by probins** [comment]
> shrinking or eliminating our dependency on the goog library (not the compiler)

if there is no longer a dependency on the library, OL users no longer have to use Closure compiler, but can use uglify or anything else they might like.

**Comment by probins** [comment]
what sort of priority does this change have? What about new PRs - does it make sense to make a rule that new PRs should not use Closure lib functions?

I strongly support this change. Apart from anything else, Closure's dependency on global namespaces looks increasingly outdated to me. Afraid I don't have much time to help out at the moment, but should have towards the end of the year.

**Comment by bjornharrtell** [comment]
I was wondering about new external (good quality) libs to replace goog.* functionality, is ol3 ready for that? Via the ext method?

**Comment by ahocevar** [comment]
> I was wondering about new external (good quality) libs to replace goog.* functionality, is ol3 ready for that? Via the ext method?

The plan was to not replace goog.* functionality with other external libs, except for what we rely on for matrix operations. Everything else should be easy to replace with ES5 functionality or minimal custom code.

That aside, the ext method works fine, and if external libraries are annotated for Closure compiler then even better.

**Comment by jirik** [comment]
Hi,
did you consider goog.module method?
https://github.com/google/closure-library/wiki/goog.module:-an-ES6-module-like-alternative-to-goog.provide
I have no experience with it yet, but maybe it can help you to achieve your goals without getting rid of the Closure Library.

**Comment by bjornharrtell** [comment]
In my opinion ES6 modules is the way to go. For bundling https://github.com/rollup/rollup is nice.

**Comment by tSte** [comment]
I've been fiddling around with latest build of Closure Compiler - using ES6 imports with Open Layers. I mean... since Closure Compiler allows mixing `goog.require/provide` with `import/export`, I've successfully removed all of the goog-stuff in my latest project. The only "real" problem I ran into was ol.Map (Map is an ES6 object), but it can be solved using `import {Map as OLMap} from 'goog:ol';`
I'd like to contribute, but I am not sure what's the community policy and/or conventions for using ES6 modules. @tschaub mentioned (~ 6 months ago) that:
> We'll leave goog.provide and goog.require until the rest of the library usage is gone.

I was unable to find any other issues relating to ES6 modules with Closure Compiler. So... what's status on this? Is the project accepting PRs for migration from Closure modules to ES6 modules?

**Comment by ahocevar** [comment]
@tSte We are definitely accepting PRs for migration to ES6 modules. In the end, it should be possible to compile OpenLayers with other standard build tools too, not just Closure compiler.

Any contributions to remove goog.* stuff from the library are much appreciated.

**Comment by probins** [comment]
@tSte could you give some more details on what you did? What, for example, does `goog:ol` consist of? As I see it, the basic issue with Closure is that its 'modules' are based on global 'namespaces', and so are incompatible with other module systems, as the whole point of those is that there are no globals. The compiler too is based on the assumption that producing a file with a lot of globals is ok. I find ol3 rather frustrating at the moment, as I would like to load modules dynamically based on demand: if a user wants to add a scaleline control or output some vectors in GPX, I should be able to just import that module, not rely on whatever has been put in some large monolithic build.

Maybe it would be a good idea to open a new issue specifically to discuss this migration.

**Comment by tSte** [comment]
@ahocevar great! I'd like to remove `goog.require/provide`... However, there is still the question about conventions. Currently, every single class is in it's namespace. Moving to import/export might produce some conflicts (like importing an OL Map class will conflict with ES6 Map). The same might happen if you try to import two (or more) classes with same name from different namespaces. What is also not clear to me is the fact whether you want to preserve namespaces or move to separate modules.

@probins well, it might not be the case anymore... [Latest GCC](https://github.com/google/closure-compiler/wiki/Releases) allows you to combine/mix module systems across the project. Simply put, you can have a script A with CommonJS, script B with ES6 and script C with goog.provide/require - and - you can compile them together. It has it's own [conventions](https://github.com/google/closure-compiler/wiki/JS-Modules) how to require e.g. `goog.provide`d symbol using ES6 module syntax. To be more specific about `goog:ol`, let's assume you are compiling OL with CommonJS and ES6 modules into one app

import {Map as olMap} from 'goog:ol'; // Importi goog.provided symbol from "ol" namespace
import myDefaultExport from './path/to/es6module'; // ES6 import
import {some, other} from './path/to/commonJS'; // Import common JS module

Note: there are [new flags](https://github.com/google/closure-compiler/wiki/Managing-Dependencies) in Closure Compiler. There are also issues in current release v20160208 related to ES6 modules an destructuring syntax (already fixed, but you have to create your own build or wait for next release). Current build does not include everything into compilation if used with `--dependency_mode=STRICT` resulting into errors because of removed interfaces. You have to explicitly import interface into each file.

Basicly, that's why I am asking about the status of ES6 modules, latest version of GCC - and ES6 usage in this project in general... Do you also accept PRs rewriting ES5 to ES6?

**Comment by bjornharrtell** [comment]
With proper ES6 modules and paths in place of namespaces it's possible to provide an entry point for a bundler that exports to global namespaces to make a backward compatible full build.

I've used this method in my own project https://github.com/bjornharrtell/jsts where I have an entry point module https://github.com/bjornharrtell/jsts/blob/master/src/jsts.js which is exported globally in https://github.com/bjornharrtell/jsts/blob/master/src/browser.js.

**Comment by tschaub** [comment]
For any 3.x releases of the library, an example like this one must still work: http://openlayers.org/en/v3.14.2/examples/simple.html

That means we have to provide a full build (and a mechanism to create a custom build) that adds everything to an `ol` object.

I was envisioning that we would not remove `goog.provide()` and `goog.require()` until the rest of `goog` is gone.  The benefit of authoring in ES6 modules will be the flexibility we provide others in using bundlers and build tools of their choosing.  As long as we have syntax that is specific to the Closure Compiler or the Closure module system, I don't see a ton of benefit of using `import` over `goog.require()`.

So, for example, if our library includes code like this:

```js
import assert from 'goog:goog.asserts'; // or whatever
```

If no other module bundler can resolve the `assert` import, then we haven't achieved much flexibility.

I don't mean to discourage all efforts in this direction.  @tSte it would be great if you can see a way we can add flexibility by moving to ES6 modules before removing other `goog` dependencies.  Or it could also just be nice to start making progress in the ES6 module direction.  Either way, we'll need to also create a way to generate builds that assign to `ol` as long as we're in the 3.x release series.

**Comment by tSte** [comment]
@tschaub that's why I am asking... well, if this cannot be done in 3.x, are there any plans for 4.0 in near future?

**Comment by bjornharrtell** [comment]
I think it makes sense to first get rid of the remaining use of `goog` until making any changes to the module system for the 3.x series.

As much as I enjoy coding in ES6 I think it might be best leave 3.x at ES5 for now. Transpiling ES6 to ES5 can in my experience be a bit unpredictable, at least for things like performance.

**Comment by tschaub** [comment]
The primary goal in removing the Closure Library dependency is to improve the user (application developer) experience.  So this is primarily about making it easier to use the library with mainstream module bundlers.

I'd say library developer experience should be second to user experience.  So, I agree with @bjornharrtell, performance takes precedence.  After moving to ES6 modules, we can talk about other ES6 features it would make sense to adopt (taking care not to introduce unnecessary `bind`ing or other performance penalties).

**Comment by probins** [comment]
> Transpiling ES6 to ES5

it's getting to the stage where the main browsers support most of ES6 natively (in fact, support more than the transpilers), so transpiling is no longer necessary except to support old browsers (like IE). Question of how much priority that has.

**Comment by gberaudo** [comment]
http://kangax.github.io/compat-table/es6/

**Comment by probins** [comment]
Strictly speaking, `goog.provide/require` are not part of removing the dependency on the Closure _library_. Once the other `goog` functions are removed, the library is no longer needed and can be deleted. Builds will continue to work, as `goog.provide/require` are built in to the compiler. So ISTM there's a second project, which is removing the dependency on the Closure _compiler_, so that apps devs can use other tools if they wish.

I just took a quick look at a simple sample file, namely https://github.com/openlayers/ol3/blob/master/src/ol/colorlike.js, to see what I would do to convert this to an ES6 module. Here's some thoughts:

1. it provides 2 symbols, one of which is only a typedef and only useful to GCC advanced. There isn't anything to export here. Some files, like `ol/transformfunction.js` only contain a typedef, so aren't really modules at all, as they don't import or export anything. As these are only for GCC, it might be better to move these into a separate directory, so they don't clutter up the actual module code. The `goog.provide` can be left in there, and picked up by GCC by those who want to continue to use it, as at present.
2. the `ol.` prefix throughout serves no real purpose in a module context, and you wouldn't use it if you were writing a plain ES6 module. Removing all these in all files, however, would be a lot of work.
3. assuming `ol/color` exports a var called `color`, you could import `color` from `ol/color` but this might clash with internal vars also called `color` (less of a problem if you use `let`), so perhaps import as some other name, like `olcolor`, would be better. Again, not difficult, but a lot of work to check and change similar things in all the files. This is similar to the problem @tSte found with `Map` clashing with reserved word `Map`.
4. on the export side, you would export a `colorlike` var. This has 2 functions, one of which is api and the other isn't. This distinction is also only useful with GCC, where 'export' means 'not obfuscated' and you can define at compile time what to 'export'. With ES6, both functions would have to be exported, so other modules could import them. You would have to find some other way of making the api/non-api distinction, if that is to continue. Looks like JSDoc has been changed to handle ES6 modules, so you would have to test whether the api docs look similar to the current ones if you move the annotations to the module export.

That's what comes to mind from looking briefly at a very small file. No doubt there are other issues in bigger ones. A lot of effort has been put into optimising the code for GCC advanced, but I'm not sure how transferable this is; have to try it and see. I don't think it's a minor task.

As I understand the GCC docs, it converts ES6 import/export into `goog.provide/require`, so at least in theory, if you make changes like these, as long as the directory structure matches the namespace names - 'ol/color' becomes 'ol.color' - you should get a build like the current one. There are though case and naming issues: 'ol/layer/vectorlayer' versus 'ol.layer.Vector'.

In the longer term, though, ISTM that the whole point of introducing modules is to get away from reliance on global variables. If you're changing to ES6 modules, which I think is the way the JS world will move in the next few years, then ISTM it would be better to have a different api based on consuming ES6 exports. `import map from 'olmap'; import * as vectorLayer from 'layer/vector'; let myMap = new map(); let myVL = new vectorLayer()` or whatever. The symbol names are defined in the `import` statements, not in the app code itself; app devs can call the imported symbols anything they like and configure the module loader to fetch 'olmap' from wherever is appropriate for that site.

So I'd see 3 different projects. I'd agree that the current priority should be to remove the dependency on the Closure _library_. Then there's a question of how to remove the dependency on the Closure _compiler_. Perhaps create a new test repo/branch with a small subset of the library, and try out various things to see what works? And finally the longer-term aim of converting to ES6 modules, and probably other ES6 stuff, as part of a new ol4, dropping support for old browsers like IE that can't handle that. With any luck, by the time you get around to this, the module loader will be implemented and there'll be experience from other projects to draw on.

**Comment by bjornharrtell** [comment]
Good points and analysis @probins, makes it clear that transitioning fully to ES6 modules is no easy task and perhaps not attainable in the 3.x series which is a bit of a bummer.

**Comment by tSte** [comment]
@probins  that's exactly what I ran into while rewriting my app from goog namespaces to ES6 modules. I was even cosidering a snippet for
```
import {Vector as VectorLayer} from 'goog:ol.layer';
import {Vector as VectorSource} from 'goog:ol.source';
```
And this is the reason why I am asking about ES6 modules PRs... It is a lot of issues to be solved, but I believe it would make sense to start working on it as soon as possible. Thus my question is the same:  Any plans for OL 4? :)

**Comment by fredj** [comment]
FYI, es6 features supported (transpiled) by the compiler:
https://github.com/google/closure-compiler/wiki/ECMAScript6

I've started to experiments this in this branch: https://github.com/openlayers/ol3/compare/master...fredj:es6

**Comment by nicholas-l** [comment]
@fredj need help? Also how easy will this be to merge in once all of the closure library is gone?

**Comment by fredj** [comment]
@nicholas-l the code in this branch can work with the closure-library.

The issue that I have right now is that the unit tests are not passing because phantomjs doesn't support es6.

The compiled code (es5) works perfectly.

**Comment by marcjansen** [comment]
> The issue that I have right now is that the unit tests are not passing because phantomjs doesn't support es6.

If we used karma, we could possibly transpile, right? From the top of my head and absolutely untested.

**Comment by fredj** [comment]
@marcjansen https://www.npmjs.com/package/karma-es6-shim ?

**Comment by probins** [comment]
but converting the code to ES6 surely isn't the priority atm. The priority is replacing goog.

**Comment by nicholas-l** [comment]
@probins think the only module that hasn't been tackled is goog.asserts but I agree.

**Comment by nicholas-l** [comment]
Although in saying that the last step would most likely change from goog.base to es6 classes, thoughts?

**Comment by bjornharrtell** [comment]
I would be wary of ES6 classes + transpilation. But perhaps you meant modules?

**Comment by fredj** [comment]
We can't use es6 classes without the modules because the class names can't have a dot, example:
```js
class ol.Map {
 ...
}
```

**Comment by probins** [comment]
I don't see any need for ES6 classes atm. I'm not sure what's planned for replacing `goog.base(this)` and `goog.inherits`, but I would have thought standard JS prototypal inheritance. Longer term, I would imagine most JS projects will replace that with classes because it's simpler, but it doesn't add any functionality.

As stated before, I think you have to get rid of the `ol.` throughout.

**Comment by marcjansen** [comment]
Jo @fredj 

> @marcjansen https://www.npmjs.com/package/karma-es6-shim ?
(From https://github.com/openlayers/ol3/issues/4128#issuecomment-206725327)

Here is an example: https://github.com/radify/karma-es6-shim-example

**Comment by nicholas-l** [comment]
Sorry what I meant was:

1. Remove closure library
2. Change to module from global namespace, either goog.module or es6 module (I assume es6 module)
3. Change to classes
All the time with transpiling (closure).

Why change to prototype if class is the end goal?

**Comment by probins** [comment]
because changing to classes is a lot more work for no real benefit.

**Comment by icholy** [comment]
Have you guys considered TypeScript? It would be a shame to throw away all that static analysis.

**Comment by probins** [comment]
I think there are other advantages to removing the Closure dependency than just increasing build options:
* few people are familiar with the Closure library so it's a barrier to potential contributors. Using standard JS code should remove that barrier.
* converting to a standard module system should mean modules can be dynamically loaded in the browser; this will be particularly attractive when a module loader is natively implemented in browsers; HTTP/2 removes much of the need for bundling. (This is my main motivation; I've been using RequireJS with OL2 for years, but frustratingly cannot use a similar system with OL3 as the current 'modules' aren't compatible with module loaders.)
* if you're loading modules individually, then you don't need a build process; you do however need to minify/compress individual modules, but advanced_optimizations is not really suitable for this, and I think most JS projects these days use UglifyJS, the equivalent of simple_optimizations. UglifyJS is however not compatible with the current library, and removing all dependencies on Closure library and compiler would enable this.
* using a standard module system also makes the library easier to extend, removing issues such as #4531.
* using a standard module system also makes it easier to create/publish add-ons; for example, at least some of the code in the examples could be converted to add-on modules which people can simply import.

**Comment by bjornharrtell** [comment]
Got curious on what remains so I did a quick search. 

The goog used most seems to be `goog.asserts` but should be simple to replace (I guess?).

Other remaining uses that I could find are `goog`, `goog.dom.NodeType`, `goog.Uri`, `goog.object`, `goog.webgl`, `goog.async.nextTick`, `goog.dom`, `goog.vec.Mat4`, `goog.color` and `goog.color.names`.

A few of the above have PRs with ongoing work, I could find are #5134, #5138 and #4194.

**Comment by ahocevar** [comment]
For `goog.asserts`, I think we will want a replacement that can be stripped out by the compiler.

For `goog.vec.Mat4`, there are concerns about bringing in a big 3rd party library that everything depends on.

The rest should be straightforward.

**Comment by probins** [comment]
see #5486 for goog.Uri. I've been looking into goog.webgl, which seems to be all constants, but I don't want to interfere with #5462. There's also goog.crypt used by strokestyle.js

**Comment by probins** [comment]
and `dom/dom.js` uses `goog.userAgent`. And see #4781 for `goog.object`.

**Comment by GaborFarkas** [comment]
Please don't worry about #5462, @probins. `goog.webgl` calls could be replaced much sooner, than that work gets finished. If there will be a merged PR with it, I will incorporate those changes in my next rebase.

**Comment by probins** [comment]
there's also a decision needed on what to do with `goog.DEBUG` - mainly used with webgl

**Comment by ahocevar** [comment]
> there's also a decision needed on what to do with goog.DEBUG - mainly used with webgl

Isn't this just a boolean flag that we could replace with `ol.DEBUG`? cc @tsauerwein

**Comment by tSte** [comment]
I have a feeling it does affect `goog.assert`-ions...

**Comment by ahocevar** [comment]
> I have a feeling it does affect goog.assert-ions...

Which we're going to replace with something non-goog* as well.

**Comment by tSte** [comment]
> Which we're going to replace with something non-goog.* as well.
True... my bad.

**Comment by probins** [comment]
> Isn't this just a boolean flag that we could replace with ol.DEBUG

I think it probably is. It's used quite a lot within Closure lib, but as that's going anyway ... Question for me is more whether it's needed at all.

**Comment by ahocevar** [comment]
> Question for me is more whether it's needed at all.

Very good question. I'm hoping that @tsauerwein can shed some light on this.

**Comment by tsauerwein** [comment]
> Question for me is more whether it's needed at all.

The code parts flagged with `goog.DEBUG` must have been useful during the initial development and I guess they will also be helpful when debugging a problem in that code. Replacing `goog.DEBUG` with `ol.DEBUG` sounds reasonable to me. When `ol.DEBUG` is set to `false`, the compiler will strip out all these parts. So keeping `ol.DEBUG` has no disadvantage in my opinion.

**Comment by probins** [comment]
it's only used a few times (37 matches) in 5 files, one of which is the auto-generated `shader.mustache`, and none of the current build configurations defines `goog.DEBUG=true`. By default, it's set to true, but what effect it has isn't documented, so it's unlikely that many people are even aware of it. If we create `ol.DEBUG`, we could give it a default of `false`, in which case most compilers/bundlers can probably exclude those bits of code, but if nobody uses it, what's the point? What can be done with it that can't be done with normal debugging techniques like setting breakpoints etc?

**Comment by probins** [comment]
other functions used widely are `goog.abstractMethod` (just throws an error) and `goog.getUid`.

**Comment by ahocevar** [comment]
Regarding type checking without Closure Compiler, I just found https://github.com/azu/babel-plugin-jsdoc-to-assert. It does not exactly replace Closure Compiler type checks because it checks at runtime rather than compile time, but we would notice type errors when running the test suite.

There is also an ongoing effort to support JSDoc type annotations in TypeScript:  Microsoft/TypeScript#7699.

**Comment by ca0v** [comment]
How about putting an incremental migration toward typescript and es6 modules on the roadmap?

**Comment by ahocevar** [comment]
We'll be fleshing this out when the Closure Library removal is complete (which should be the case in a few days). ES2015 modules has higher priority for now.

**Comment by probins** [comment]
the final stage in this is telling closure-util to stop using the Closure lib. AFAICS it isn't documented anywhere, but it looks like there is a [closure option](https://github.com/openlayers/closure-util/blob/master/lib/manager.js#L99-L105) which could be set in tasks/build.js. The postinstall task in `package.json` would also need to be changed to `update-compiler`. Anything else?

**Comment by adube** [comment]
In my projects, I use closure-util to build ol3 along with custom code, i.e. I use the tasks/build.js script of OL3 itself to run the builds.  In them, I also still use the Closure lib.  Is there a way this could be an option in tasks/build.js for people to still be able to run the build script with or without Closure lib ?  The default behavior could be "without".

Thanks.

**Comment by probins** [comment]
what's left apart from DEBUG (editing out some mentions of 'goog' in comments)
```
cd src/ol; grep "require('goog" -nr .; grep 'goog\.' -nr . | grep -E 'goog\.provide|goog\.require|DEBUG' -v
./source/tileutfgridsource.js:3:goog.require('goog.async.nextTick');
./map.js:8:goog.require('goog.async.nextTick');
./dom/dom.js:3:goog.require('goog.userAgent');
./style/strokestyle.js:3:goog.require('goog.crypt');
./style/strokestyle.js:4:goog.require('goog.crypt.Md5');
./format/kmlformat.js:8:goog.require('goog.object');
./webgl/shader.mustache:19:goog.addSingletonGetter({{className}}Fragment);
./webgl/shader.mustache:54:goog.addSingletonGetter({{className}}Vertex);
./render/webgl/webglimagedefaultshader.js:19:goog.addSingletonGetter(ol.render.webgl.imagereplay.shader.DefaultFragment);
./render/webgl/webglimagedefaultshader.js:54:goog.addSingletonGetter(ol.render.webgl.imagereplay.shader.DefaultVertex);
./events/condition.js:77:      !(goog.userAgent.WEBKIT && ol.has.MAC && originalEvent.ctrlKey);
./events/condition.js:208:  goog.asserts.assertInstanceof(mapBrowserEvent, ol.MapBrowserPointerEvent,
./renderer/webgl/webgltilelayershader.js:19:goog.addSingletonGetter(ol.renderer.webgl.tilelayer.shader.Fragment);
./renderer/webgl/webgltilelayershader.js:54:goog.addSingletonGetter(ol.renderer.webgl.tilelayer.shader.Vertex);
./renderer/webgl/webglmapdefaultshader.js:19:goog.addSingletonGetter(ol.renderer.webgl.map.shader.DefaultFragment);
./renderer/webgl/webglmapdefaultshader.js:54:goog.addSingletonGetter(ol.renderer.webgl.map.shader.DefaultVertex);
./source/tileutfgridsource.js:141:      goog.async.nextTick(function() {
./source/tileutfgridsource.js:393:      goog.async.nextTick(function() {
./map.js:1318:  goog.async.nextTick(this.handlePostRender, this);
./dom/dom.js:118:  if (goog.userAgent.IE && goog.userAgent.isVersionOrHigher('9.0')) {
./style/strokestyle.js:224:    var md5 = new goog.crypt.Md5();
./style/strokestyle.js:226:    this.checksum_ = goog.crypt.byteArrayToString(md5.digest());
./style/iconstyle.js:663:goog.addSingletonGetter(ol.style.IconImageCache);
./format/kmlformat.js:316:        (goog.object.clone(foundStyle.getText()));
```

**Comment by ahocevar** [comment]
> the final stage in this is telling closure-util to stop using the Closure lib. AFAICS it isn't documented anywhere, but it looks like there is a closure option which could be set in tasks/build.js. The postinstall task in package.json would also need to be changed to update-compiler.

We'll still need `goog/base.js` for the code generated by `tasks/generate-exports.js`. If we'd replace that with `@export` annotations for everything we currently have marked `@api`, we'd still need to get in #5679 first. Why? Because the code generated by `tasks/generate-exports.js` adds a `goog.require` for everything. When that is gone, the compiler will complain about missing requires in `src/ol/typedefs.js`.

**Comment by probins** [comment]
I think goog.base is still needed for the debug build too, right?

**Comment by ahocevar** [comment]
> I think goog.base is still needed for the debug build too, right?

Right.

**Comment by probins** [comment]
although there wouldn't be any need to keep it updated - just keep a copy somewhere

**Comment by probins** [comment]
and closure-util uses base.js for working out dependencies (it throws an error if it can't find base.js) - though this stage could be skipped by passing the compiler the src globs instead the list of files created by getDependencies, and letting the compiler work out which files are needed

**Comment by gberaudo** [comment]
For `goog.module` we also need some functions from base.js.

**Comment by ahocevar** [comment]
I have no problem keeping base.js. It gets completely stripped out of the Closure Compiler build anyway, and it won't make a difference for an alternative build process as outlined in https://github.com/openlayers/ol3/issues/5679#issue-169166581.

**Comment by probins** [comment]
> replace all goog.exportSymbol and good.exportProperty with `@export` annotations (not sure if this would work)

don't think it would: see https://developers.google.com/closure/compiler/docs/js-for-compiler#tags "Code that uses the `@export` annotation must either include closure/base.js, or define both goog.exportSymbol and goog.exportProperty with the same method signature in their own codebase." I tried a couple of experiments using this branch. Removing base.js from standard full build leads to lots of undefined property errors. Removing it from my custom build, which uses default error/warning, compiles without error but leaves all the goog.exportSymbol and goog.exportProperty in the compiled output, which isn't much use.

> keeping base.js. It gets completely stripped out of the Closure Compiler build

not entirely true. There is an issue with putting the build in an ES6 module wrapper. At the front of the build, it puts `var aa=this;` which won't work in ES6 module, where top-level `this` is undefined. While the code was still using Closure lib, this had to be edited to `var aa=window;` so the Closure functions worked. I was hoping that `var aa=this;` would disappear now that Closure is gone, but it looks like it's `base.js` that's adding it, as it's still in there (probably from `goog.global = this;`). As long as you're using the namespace option, it doesn't matter because it uses the namespace and not `aa`. I'll add a note to tasks/readme about this.

**Comment by probins** [comment]
I think this should be kept open until Closure lib is removed from the build process (or alternatively a new issue/PR is opened for that)
